pkg/bpf/map_linux.go

```diff
@@ -675,16 +677,26 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error
 	bpfNextKeyPtr := uintptr(unsafe.Pointer(&bpfNextKey))
 	bpfNextKeySize := unsafe.Sizeof(bpfNextKey)
 
-	for stats.Lookup = 1; stats.Lookup <= stats.MaxEntries; stats.Lookup++ {
-		// currentKey was returned by GetNextKey() so we know it existed in the map, but it may have been
-		// deleted by a concurrent map operation. If currentKey is no longer in the map, nextKey will be
-		// the first key in the map again. Use the nextKey only if we still find currentKey in the Lookup()
-		// after the GetNextKey() call, this way we know nextKey is NOT the first key in the map.
-		nextKeyValid := GetNextKeyFromPointers(m.fd, bpfNextKeyPtr, bpfNextKeySize)
+	// maxBacktrack is an upper bound limit to prevent backtracking forever
+	// when iterating over the map's elements (the map might be concurrently
+	// updated while being iterated)
+	maxBacktrack := 4 * stats.MaxEntries
+
+	// this loop stops when all elements have been iterated
+	// (GetNextKeyFromPointers returns io.EOF) OR, in order to avoid hanging if
+	// the map is continuously updated, when maxBacktrack has been reached
+	for ; stats.LookupFailed < maxBacktrack; stats.Lookup++ {
```

Comment:
Well, I guess it doesn't sound quite so bad in these terms; today we would iterate as many as 2 million syscalls for a 1-million entry map. But the above is with a pretty low probability of resetting the full dump, and it can get significantly longer if that probability rises.

Reply:
Now set to 4 * stats.Lookups rather than failures.